Conditional deps for plotting

svilupp opened the pull request:
Quick draft tackling https://github.com/mschauer/CausalInference.jl/issues/73

Goals: 
- remove direct dependency on TikzGraphs
- introduce a (conditional) plotting method for GraphRecipes
- and, hopefully, a simple text-based output

It needs some thinking around the function names and re-use. 

Done so far:
- added conditional dependencies ala new Package Extensions
- separated preparation of a graph and plotting, as the former can be re-used for many backends
- added a quick draft of text output with TabularDisplay.jl (which add another direct dependency but it's fairly light and stable)

Loud thoughts:
- tabular display looks awful for small DAGs
- how to forward graphs to the right plotting backend/method/package (ie, using the same function but forwarding to the right method when it's available)
- (related) how to load Requires in if-else form to choose one active backend // or how to avoid method redefinition with potentially better methods in Requires.jl

mschauer commented:
I actually quite like what is already there. The additional dependency is no problem.

svilupp commented:
I'm planning to tidy it up tonight. I'll probably implement the below proposal.

Would you have any thoughts on how to handle multiple backends, ie, what is the `plot` method to use without redefinitions? 
I couldn't find any examples of how to do if-else with Requires.jl and I don't want to meddle with the user's multimedia display stack.

My proposal:
- set a package-level const PLOT_BACKEND=[]
- then lazy load code
    - try Tikz and if PLOT_BACKEND is empty, push tikz-based method in
    - try GraphRecipes and if PLOT_BACKEND is empty, push graphplot-based method in
- if PLOT_BACKEND is still empty, push the text-based method in
- when plotting, grab the method in PLOT_BACKEND

It feels unnecessarily complex, so perhaps you have some better ideas.

___
Separately, the docstrings are fairly plain with no arguments and examples. I adhered to the same style for consistency, but I could also add some examples and comment on the arguments for the methods I touch. WDYT?

mschauer commented on src/pc.jl:
```diff
@@ -279,11 +278,11 @@ function pcalg(t, p::Float64, test::typeof(gausscitest); kwargs...)
+    N = size(X, 1)
     C = Statistics.cor(X)
-    return pcalg(n, gausscitest, (C,N), quantile(Normal(), 1-p/2); kwargs...)
+    return pcalg(n, gausscitest, (C, N), quantile(Normal(), 1 - p / 2); kwargs...)
```
```suggestion
    return pcalg(n, gausscitest, (C, N), quantile(Normal(), 1 - p/2); kwargs...)
```
I like it if whitespace use clarifies precedence.

svilupp replied:
This wasn't intentional - it was the "format on save" in VS Code. I can nuke all these changes when everything is ready, but it would be even better to set a consistent formatting for the package with https://github.com/domluna/JuliaFormatter.jl 

Would you have a preferred style? Blue/SciML?

svilupp replied:
Okay, I'll add a formatter .toml file specifying [SciML style](https://github.com/SciML/SciMLStyle) and run the whole code base through it.

As for the operator precedence-based spacing, I cannot find the corresponding option [here](https://domluna.github.io/JuliaFormatter.jl/stable/) - do you know which one it would be?

mschauer replied:
(The rationale being that whitespace around multiplications is less bad than what we have now ;-) )

mschauer commented:
I think it's absolutely okay to have functions with different names `plot_pc_graph_backend` if that helps sorting things.

svilupp commented:
Okay, a quick update: 
- I've added plotting methods with a suffix for each backend
- conditional loading (tested for GraphRecipes+Plots, cannot test it for TikzGraphs as I cannot install it)
- moved empty functions and their docstrings into the main source, so users know how to access the plots
- added some light tests for the plotting prep and text-based backend

mschauer commented:
Once this converges: there are some examples and doc files that use plotting.

mschauer commented on src/graphs.jl:
```diff
@@ -35,4 +36,41 @@ function has_a_path(g::AbstractGraph{T}, U::Vector, V::Vector, exclude_vertices:
+"""
+    graph_to_text(g::AbstractGraph, node_labels::AbstractVector{<:AbstractString}=[]; edge_styles::AbstractDict=Dict())
+
+print out a graph `g` as a table of edges labeled by`node_labels`
```
```suggestion
Print out a graph `g` as a table of edges labeled by`node_labels`.
```

mschauer commented on src/pc.jl:
```diff
@@ -304,36 +303,69 @@ function pcalg(t, p::Float64, test::typeof(cmitest); kwargs...)
+"""
+    plot_pc_graph_text(g::AbstractGraph, node_labels::AbstractVector{<:AbstractString}=String[])
+
+plot the output of the PG algorithm (Text-based output)
```
```suggestion
Plot the output of the PG algorithm (text-based output).
```

mschauer commented on src/pc.jl:
```diff
@@ -304,36 +303,69 @@ function pcalg(t, p::Float64, test::typeof(cmitest); kwargs...)
+"""
+    plot_pc_graph_recipes(g, node_labels::AbstractVector{<:AbstractString}=String[])
+
+plot the output of the PC algorithm (GraphRecipes backend)
```
```suggestion
Plot the output of the PC algorithm (GraphRecipes backend).
```

mschauer commented on src/pc.jl:
```diff
@@ -304,36 +303,69 @@ function pcalg(t, p::Float64, test::typeof(cmitest); kwargs...)
+"""
+    CausalInference.plot_pc_graph_tikz(g, node_labels::AbstractVector{<:AbstractString}=String[])
+
+plot the output of the PC algorithm (TikzGraphs backend)
```
```suggestion
Plot the output of the PC algorithm (TikzGraphs backend).
```

svilupp commented:
Changes made so far:
- reflected your docstring suggestions (updated also other docstrings in pc.jl, fci.jl, graphs.jl scripts for consistency)
- updated examples and docs .md files (added also relevant dependencies, so each is separately executable) + added shout out to new plotting methods if users have problems with Tikz
- updated docs/project.toml to be able to execute the docs (it had only Documenter dep? Not sure how it worked)

I think it should be ready for review. Let me know if you have any suggestions.

mschauer commented:
Thank you, that is very nice. What is left to do?

svilupp commented:
So I'm unable to check the Tikz-based examples and functionality, so it would be awesome if you double check that I did everything right.

Otherwise, it should be ready.

mschauer commented:
Great, just checked it. Thank you very much.
